File: PyRTAlib/RTAInterface/RTA_DL_DB.py
# ==========================================================================
#
# Copyright (C) 2018 INAF - OAS Bologna
# Author: Leonardo Baroncelli <leonardo.baroncelli26@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# ==========================================================================

from abc import ABC, abstractmethod
import queue
import threading

from ..DBConnectors import RedisDBConnector, MySqlDBConnector, RedisDBConnectorBASIC
from ..Utils import Config


class RTA_DL_DB(ABC):

    def __init__(self, database, configFilePath = '', **kwargs):
        super().__init__()

        if database != 'mysql' and database != 'redis' and database != 'redis-basic':
            print("[RTA_DL_DB] Database '{}' is not supported. Supported databases: \n- {}\n- {}".format(database,'mysql','redis'))
            exit()

        self.config = Config(configFilePath) # singleton config object

        # Synchronous (master thread) execution /\____/\____/\____/\____/\____/\
        if self.config.get('General', 'numberofthreads', 'int') == 1:
            self.dbConnector = self.getConnector(database, configFilePath)
            self.dbConnector.connect()


        # Multi threading mode /\____/\____/\____/\____/\____/\____/\____/\____/\
        else:

            """
            The queue module implements multi-producer, multi-consumer queues. It is
            especially useful in threaded programming when information must be
            exchanged safely between multiple threads.
            """
            self.eventQueue = queue.Queue(maxsize=-1) # queue size is infinite.


            # batchsize = 1 -> threadsnumber = 1 ?
            #if self.config.get('General', 'batchsize', 'int') == 1):
            #    self.config.set('General', 'numberofthreads', 1)

            """
                Variables to stop the threads
            """
            self.run = []
            for i in range(self.config.get('General','numberofthreads', 'int')):
                self.run.append(True)

            """
                Running the threads. Each thread has its own db connector.
            """
            # self.threads = []
            for i in range(self.config.get('General','numberofthreads', 'int')):

                if self.config.get('General','debug') == 'yes':
                    print("[RTA_DL_DB] Starting new thread!")

                dbConnector = self.getConnector(database, configFilePath)
                t = threading.Thread(target=self.consumeQueue, args=(i, dbConnector))
                # self.threads.append(t)
                t.start()


    def __enter__(self):
        return self


    def __exit__(self, exc_type, exc_value, traceback):
        self.close()


    @abstractmethod
    def insertEvent(self, *args):
        pass


    def _insertEvent(self, event):
        # Single thread
        if self.config.get('General', 'numberofthreads', 'int') == 1:
            self.dbConnector.insertData(self.config.get('General','evt3modelname'), event.getData())
        # Multi threading mode
        else:
            self.eventQueue.put_nowait(event)

    def readNewData(self):
        try:
            return self.eventQueue.get_nowait()
        except queue.Empty:
            return None

    def getConnector(self, database, configFilePath):
        if database == 'mysql':
            return MySqlDBConnector(configFilePath)
        elif database == 'redis':
            return RedisDBConnector(configFilePath)
        elif database == 'redis-basic':
            return RedisDBConnectorBASIC(configFilePath)


    def consumeQueue(self, threadId, dbConnector):
        if self.config.get('General','debug') == 'yes':
            print('-->[RTA_DL_DB thread: {} ] Starting..'.format(threadId))

        dbConnector.connect()

        while self.run[threadId]:
            event = self.readNewData()

            if event is not None:
                if not dbConnector.insertData(self.config.get('General','evt3modelname'), event.getData()):
                    print("-->[RTA_DL_DB thread: {} ] DBconnector insert data error. ".format(threadId))
                    break;

                elif self.config.get('General','debug') == 'yes':
                    print("-->[RTA_DL_DB thread: {} ] Data inserted: {} ".format(threadId, event.getData()))

        if self.config.get('General','debug') == 'yes':
            print("-->[RTA_DL_DB thread: {} ] Job finished..")
        dbConnector.close()


    def close(self):

        # TODO  Should close when all threads finish their job.

        if self.config.get('General','debug') == 'yes':
            print('[RTA_DL_DB] Stopping all threads on close()..')

        if self.config.get('General', 'numberofthreads', 'int') > 1:
            for i in range(self.config.get('General', 'numberofthreads', 'int')):
                self.run[i] = False
        else:
            self.dbConnector.close()

    def getThreads(self):
        return threading.enumerate()

File: CxxRTAlib/CxxRTAlib/RTAInterface/src_cpp/RTA_DL_DB.cpp
/*
 ==========================================================================

 Copyright (C) 2018 Giancarlo Zollino
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ==========================================================================
*/


#include"RTA_DL_DB.hpp"

RTA_DL_DB::RTA_DL_DB(string database, string configFilePath){

  if (database != "mysql" && database != "redis-basic"){

    cout << "[RTA_DL_DB] Error! Database '{ " << database << " }' is not supported. Supported databases: \n- { mysql }\n- { redis-basic }" << endl;
    exit(EXIT_FAILURE);

  }

  #ifdef DEBUG
  cout << "RTA_DL_DB configFilePath: " << configFilePath << endl;
  #endif

  config = Config::getIstance(configFilePath);

  numberofthreads = config->file["General"]["numberofthreads"].getInt();
  modelname = config->file["General"]["modelname"].getString();
  DTRinChannel = config->file["Dtr"]["inputchannel"].getString();
  Mutex* mux = Mutex::getIstance();

  //countevts = 0;

  // DTR Configuratoion
  DTRactive = config->file["Dtr"]["active"].getString();
  if ( DTRactive.compare("yes") == 0 ){
    BoolDTRactive = true;
    cout << "DTR attivo" << endl;
    redisPub = make_shared<RedisPublisher>(configFilePath);
  }else{
    BoolDTRactive = false;
  }


  if( numberofthreads == 0 ) {

    if( BoolDTRactive) {
        redisPub = make_shared<RedisPublisher>(configFilePath);
    }

    // Synchronous (master thread) execution
    dbConnector = getConnector(0, database, configFilePath);

    if( dbConnector->connect(mux) == false ) {

      cout << "CXX_RTA_DL_X_DB Connection Error!" << endl;

      exit(EXIT_FAILURE);

    }

  }else if( numberofthreads >= 1) {

    // Multi threading mode
    eventBuffer = new CTABuffer("rta_dl_buffer",1000);


    // for( int i = 0; i < numberofthreads; i++ ) {
    //  // ThreadStatistic * ts = new ThreadStatistic(i);
    //  // thread_statistics_array.push_back(ts);
    // }

    for( int i = 0; i < numberofthreads; i++ ) {

      shared_ptr<RedisPublisher> redisPubThread = NULL;

      if(BoolDTRactive){
          redisPubThread = make_shared<RedisPublisher>(configFilePath);
      }

      dbConnector = getConnector(i, database, configFilePath);

      // ThreadStatistic * ts = thread_statistics_array[i];
      // ts->printThreadId();

      auto t = make_shared<RTAThread>(i, mux, modelname, dbConnector, eventBuffer, redisPubThread, BoolDTRactive, DTRinChannel);
      thread_array.push_back(t);

    }

  }



}

void RTA_DL_DB::start() {


  if( numberofthreads == 0 ) {

  }else {
    for( int i = 0; i < numberofthreads; i++ ) {
      // #ifdef DEBUG
      cout << "[RTA_DL_DB] Starting thread: " <<i << endl;
      // #endif
      thread_array[i]->start();
    }
  }

  starttime = Clock::now();

}


shared_ptr<DBConnector> RTA_DL_DB::getConnector(int id,string databaseEngine, string configFilePath) {

  if(databaseEngine == "mysql"){

    auto mySqlDBConnector = make_shared<MySqlDBConnector>(id,configFilePath);

    return mySqlDBConnector;

  } else if(databaseEngine == "redis-basic"){

    auto redisDBConnector = make_shared<RedisDBConnector>(id,configFilePath);

    return redisDBConnector;

  }

}

bool RTA_DL_DB::_insertEvent( EVTbase *event ) {

  // countevts++;

  string modelname = config->file["General"]["modelname"].getString();



  if( numberofthreads == 0 ) {

    map < string, string > eventData =  event->getData();


    // #ifdef DEBUG
    // cout << "DEBUG event print in RTA_DL_DB" << " \n" <<
    //         eventData["eventidfits"] << " \n" <<
    //         eventData["timerealtt"] << " \n"<<
    //         eventData["ra_deg"] << " \n"<<
    //         eventData["dec_deg"] << " \n"<<
    //         eventData["energy"] << " \n"<<
    //         eventData["detx"] << " \n"<<
    //         eventData["dety"] << " \n"<<
    //         eventData["observationid"] << " \n"<<
    //         eventData["datarepositoryid"] << " \n"<<
    //         eventData["status"] << " \n"<<
    //         eventData["mcid"] << " \n"<<
    //         eventData["insert_time"]  << endl;
    // #endif

    // Synchronous (master thread)
    delete event;
    bool ok = dbConnector->insertData(modelname, eventData);

    #ifdef DEBUG
    cout << "BoolDTRactive value: " << BoolDTRactive << endl;
    #endif

    // Transform data for visualization and notify GUIs
    if(BoolDTRactive){
      redisPub->Publish(DTRinChannel, eventData);
    }

    return ok;


  }else if( numberofthreads >= 1 ){

    // Multi threading mode
    eventBuffer->put(event);
    #ifdef DEBUG
    cout<< "\n[RTA_DL_DB] Event inserted in buffer" << endl;
    #endif

    return 1;

  }

}


bool RTA_DL_DB::waitAndClose() {

  #ifdef DEBUG
  cout << "\n[RTA_DL_DB] waitAndClose" << endl;
  #endif

  if( numberofthreads >= 1 ) {

    for(int i=0; i < numberofthreads; i++ ) {
      EVTbase * stopEvent = new EVTbase(true);
      _insertEvent(stopEvent);
    }

    for(int i=0; i < numberofthreads; i++ ) {
      thread_array[i]->join();
      //endtime = std::chrono::system_clock::now();
      //std::chrono::duration<double> diff = endtime - starttime;
      //std::cout << "Took time: "<< diff.count() << " s\n";

      #ifdef DEBUG
      cout << "[RTA_DL_DB] thread "<< i << " joined" << endl;
      #endif
    }

    // cout << "[RTA_DL_DB] Collecting statistics.. " << endl;
    // for(int i=0; i < numberofthreads; i++ ) {
    //   // thread_statistics_array[i]->getStatistics();
    // }

  }else{

    //endtime = std::chrono::system_clock::now();
    //std::chrono::duration<double> diff = endtime -starttime;
    //std::cout << "Took time: "<< diff.count() << " s\n";
    auto elapsed = Clock::now() - starttime;
    cout <<"[RTAlib] took: " << std::chrono::duration_cast<std::chrono::milliseconds>(elapsed).count() << "ms" << endl;

    if ( dbConnector->disconnect() == true ) {
      cout << "[RTA_DL_DB] disconnected" << endl;
      return true;
    }else {
      return false;
    }
  }
}

int RTA_DL_DB::getNumberOfThreads() {
  return numberofthreads;
}
